lib/engine/game/g_1862/game.rb

```diff
@@ -297,9 +405,99 @@ def init_companies(players)
 
         def setup; end
 
+        def setup_preround
+          @double_parliament = true
+
+          # randomize order of corporations, then remove some based on player count
+          @offer_order = @corporations.sort_by { rand }
+          num_removed = case @players.size
+                        when 8
+                          2
+                        when 7
+                          3
+                        else
+                          4
+                        end
+          removed = @offer_order.take(num_removed)
+          removed.each do |corp|
+            @offer_order.delete(corp)
+            @corporations.delete(corp)
+            remove_reservation(corp)
+            @companies.delete(@companies.find { |c| c.id == corp.id })
+
+            @log << "Removing #{corp.name} from game"
+          end
+
+          @chartered = {}
+
+          # randomize and distribute train permits
+          permit_list = %i[
+            freight
+            freight
+            freight
+            freight
+            freight
+            freight
+            express
+            express
+            express
+            express
+            express
+            local
+            local
+            local
+            local
+            local
+          ]
+          permit_list.concat(%i[express local]) if @players.size > 6
+          rand_permits = permit_list.sort_by { rand }
+          @permits = Hash.new { |h, k| h[k] = [] }
+          @corporations.each_with_index { |corp, idx| @permits[corp] << rand_permits[idx] }
+
+          # record what phases corp become available
+          @starting_phase = {}
+          @offer_order.each { |c| @starting_phase[c] = 'A' }
+          @offer_order.reverse.take(8).each { |c| @starting_phase[c] = 'B' }
+          @offer_order.reverse.take(4).each { |c| @starting_phase[c] = 'C' }
+        end
+
+        def remove_reservation(corporation)
+          hexes.each do |hex|
+            hex.tile.cities.each do |city|
+              city.reservations.delete(corporation) if city.reserved_by?(corporation)
+            end
+          end
+        end
+
+        def ipoable_railroads
+          available_railroads.reject(&:ipoed)
+        end
+
+        def available_railroads
```

Comment:
i'm not sure about mixing corporation and railroads in terminology here.

Reply:
I wasn't either, but I didn't want to have both available_corporations and corporation_available?
I'll see if I can find a less offensive name.